### ISSUE TITLE
Receive error when trying to access service with impersonated creds

### ISSUE THREAD

**Issue by jembishop** [pythongssapi/python-gssapi#164]
### What went wrong?
Hi I am having a problem with the impersonation feature. I am not very experienced with kerberos so I might be doing something stupid. I am not using the real realm and usernames in this example, with logging set to info. 

### How do we reproduce?

As I understand it to get your creds you should do:
```code
import gssapi
creds = gssapi.Credentials.acquire(gssapi.Name('jemb'),usage="initiate",store={"ccache":"/tmp/krb5cc_1001747_hAPX7qzhxX"}).creds
```
I has to use this cred cache as the one it defaulted to had expired creds. 

I am then using:
```code
imp_creds=creds.impersonate(gssapi.Name('other-user'))
imp_auth = HTTPSPNEGOAuth(mutual_authentication=requests_gssapi.DISABLED,creds=imp_creds)
```
to get an impersonated auth.
This seems to work:

```code 
>>> imp_creds.inquire()
InquireCredResult(name=Name(b'other-user@EXAMPLE.NET', <OID 1.2.840.113554.1.2.2.1>), lifetime=5665, usage='initiate', mechs={<OID 1.2.840.113554.1.2.2>})
```

However when I go  to use this to access an internal service on the network I get:

```code
>>> requests.get("http://exampleservice/hello", auth=imp_auth, verify=False)
ERROR:requests_gssapi.gssapi_:Major (851968): Unspecified GSS failure.  Minor code may provide more information, Minor (2529639053): Matching credential not found
Traceback (most recent call last):
  File "/home/jemb/amazingvenv/lib/python3.6/site-packages/requests_gssapi/gssapi_.py", line 148, in generate_request_header
    _negotiate_value(response))
  File "<decorator-gen-15>", line 2, in step
  File "/home/jemb/amazingvenv/lib/python3.6/site-packages/gssapi/_utils.py", line 167, in check_last_err
    return func(self, *args, **kwargs)
  File "<decorator-gen-5>", line 2, in step
  File "/home/jemb/amazingvenv/lib/python3.6/site-packages/gssapi/_utils.py", line 127, in catch_and_return_token
    return func(self, *args, **kwargs)
  File "/home/jemb/amazingvenv/lib/python3.6/site-packages/gssapi/sec_contexts.py", line 521, in step
    return self._initiator_step(token=token)
  File "/home/jemb/amazingvenv/lib/python3.6/site-packages/gssapi/sec_contexts.py", line 542, in _initiator_step
    token)
  File "gssapi/raw/sec_contexts.pyx", line 244, in gssapi.raw.sec_contexts.init_sec_context
gssapi.raw.misc.GSSError: Major (851968): Unspecified GSS failure.  Minor code may provide more information, Minor (2529639053): Matching credential not found
```
Accessing with my own creds works as expected though. 
I should be set up for constrained delegation to the service I am trying to access. Any help would be appreciated. 

### Component versions (python-gssapi, Kerberos, OS / distro, etc.)
pip installed gssapi==1.5.1
Windows Kerberos.
RHEL.

**Comment by frozencemetery** [comment]
Please show all of the code you're using.  You're missing a definition for imp_auth if we just take the lines you've shown here.

I'm also curious what you're trying to do with impersonation.  If you're new to GSSAPI/kerberos, you may find https://ssimo.org/blog/id_016.html helpful

**Comment by jembishop** [comment]
Hi frozencemetery, thanks for your quick response.As I mentioned above I defined:
```code
imp_creds=creds.impersonate(gssapi.Name('other-user'))
imp_auth = HTTPSPNEGOAuth(mutual_authentication=requests_gssapi.DISABLED,creds=imp_creds)
```

**Comment by jembishop** [comment]
I have managed to get this working with the S4U2Proxy delegation feature.

**Comment by ptulpen** [comment]
@jembishop : can you please post how you solved it with the S4U2Proxy delegation feature ?
(We currently facing the same issue with an flask app)

**Comment by jembishop** [comment]
@ptulpen 
- Make sure your user has the necessary delegation perms set up in the KDC.
- Authenticate the user as normal, ie wait for `server_ctx.complete==True`
- After that the delegated creds should be available from the server_ctx as `server_ctx.delegated_creds`.
- To authenticate using these delegated creds do something like 
```python
import requests_gssapi
import requests

# do authentication
...
delegated_creds = server_ctx.delegated_creds

r = requests.get(my_url, auth=requests_gssapi.HTTPSPNEGOAuth(creds=delegated_creds)
...
```

The delegated credentials will then be used instead of the default credentials.

**Comment by ptulpen** [comment]
Hello jembishop,
thanks a lot for your answer. 
Unfortunately I am still stuck.
What I try to achieve is to combine this solution: https://fy.blackhats.net.au/blog/html/2015/11/26/python_gssapi_with_flask_and_s4u2proxy.html
with your answer.
(basicially recieve a TGT and create a new kerberized connection)
I already now recognized that the pip packages gssapi and python-gssapi are not equal or compatible *doh*

So long story short:
could you please give me an example of
` # do authentication` ?

**Comment by jembishop** [comment]
here is an example of how to do this.
``` python 
import base64
from functools import wraps
import logging

import gssapi

from flask import Response, request, make_response

logger = logging.getLogger(__name__)

class KerberosAuthenticator:
    """Provides a decorator for kerberos auth in flask"""
    def __init__(self):
        self.server_creds = gssapi.Credentials.acquire().creds

    def authenticate(self, func):
        """Decorator for making flask view function require kerberos auth"""
        @wraps(func)
        def decorated(*args, **kwargs):
            client_auth_header = request.headers.get("Authorization", "")
            if (
                not client_auth_header.startswith("Negotiate ")
                or len(client_auth_header.split()) != 2
            ):
                logger.debug("No token returning 401")
                return Response("Unauthorized", 401, {"WWW-Authenticate": "Negotiate"})
            b64_client_token = client_auth_header.split()[1]
            client_token = base64.b64decode(b64_client_token)
            logger.debug("Attempting to validate token...")
            server_context = gssapi.SecurityContext(
                creds=self.server_creds, usage="accept"
            )
            try:
                server_token = server_context.step(client_token)
                if server_token:
                    b64_server_token = base64.b64encode(server_token).decode("ascii")
                else:
                    b64_server_token = ""
                logger.debug("Server token: {}".format(b64_server_token))
                if not server_context.complete:
                    logger.debug("Authentication incomplete sending server token..")
                    return Response(
                        "Unauthorized",
                        401,
                        {"WWW-Authenticate": "Negotiate {}".format(b64_server_token)},
                    )
            except gssapi.exceptions.GSSError as e:
                logger.info("Authentication failed")
                logger.debug("GSSAPI error: {}".format(e))
                return Response("Unauthorized", 401)
            logger.info("Authenticated {}".format(server_context.initiator_name))
            request.kerb_ctx = server_context
            response = make_response(func(*args, **kwargs))
            response.headers["WWW-Authenticate"] = "Negotiate {}".format(
                b64_server_token
            )
            return response
        return decorated
```
Then you can use the decorator provided to authenticate.
``` python
from flask import Flask, request

app = Flask(__name__)
kauth = KerberosAuthenticator()


@kauth.authenticate
def view():
    r = requests.get(
        my_url,
        auth=requests_gssapi.HTTPSPNEGOAuth(creds=request.kerb_ctx.delegated_creds),
    )
    return f"hello you are {request.kerb_ctx.initiator_name}"

```
You can see that I have added the server context to the request context global. From this you can get the initiator name and delegated creds.

**Comment by ptulpen** [comment]
well, I am getting closer ;)
What I get now is that  request.kerb_ctx.initiator_name is filled correctly, but request.kerb_ctx.delegated_creds is empty.  (I expanded your "hello you are... " answer) 
Therefore, I see that the request  to my_url is made, but unauthenticated.

**Comment by jembishop** [comment]
You may need to set the `delegate=True` flag when you make a request to the app.
``` python 
r = requests.get(your_service_url, requests_gssapi.HTTPSPNEGOAuth(delegate=True))
```

**Comment by ptulpen** [comment]
now I get a kerberised connection, but it is not with the delegated user, but with the one who has the active ticket when it was started 
so when I do an kinit myadmin ahead of starting, myadmin creates the request
Even when using kinit -kt /etc/http.keytab   [myspn], the spn gets logged as user in the backend.
So I think there must be another point where to hint him to use delegated credentials instead of the local one

**Comment by jembishop** [comment]
What is request.kerb_ctx.delegated_creds now?

**Comment by ptulpen** [comment]
I added 
```
            server_context = gssapi.SecurityContext(
                creds=self.server_creds, usage="accept"
            )
            print("server context delegated creds")
            print(server_context.delegated_creds)
```
wich gives: 
server context delegated creds
None


and in the view function 
`  
 print("delegated creds in request")
    print(request.kerb_ctx.delegated_creds)
    r = requests.get(
`
which then outputs this: 
delegated creds in request
<gssapi.creds.Credentials object at 0x7fc4248160d0>

**Comment by jembishop** [comment]
Hmm ok so the delegated creds are there but you are saying they are not for the requesting user?

Maybe try setting ‘logging.basicConfig(level=“DEBUG”)’ 
And set ‘KRB5_TRACE=/dev/stdout”. And read the logs to find some more clues?

**Comment by ptulpen** [comment]
What I get now is the following output
the user there is the correct one, so I dont see why he assumes matching credential is not found

> 
> 11512] 1562863977.545772: Getting credentials MYUSER@EXAMPLE.NET -> HTTP/servername.EXAMPLE.NET@EXAMPLE.NET using ccache MEMORY:jDP7v1E
> [11512] 1562863977.545773: Retrieving MYUSER@EXAMPLE.NET -> HTTP/servername.EXAMPLE.NET@EXAMPLE.NET from MEMORY:jDP7v1E with result: -1765328243/Matching credential not found
> [11512] 1562863977.545774: Retrieving MYUSER@EXAMPLE.NET -> HTTP/servername.EXAMPLE.NET@EXAMPLE.NET from MEMORY:jDP7v1E with result: -1765328243/Matching credential not found
> ERROR:requests_gssapi.gssapi_:generate_request_header(): stepping context failed:
> Traceback (most recent call last):
>   File "/opt/miniconda3/lib/python3.7/site-packages/requests_gssapi/gssapi_.py", line 150, in generate_request_header
>     _negotiate_value(response))
>   File "</opt/miniconda3/lib/python3.7/site-packages/decorator.py:decorator-gen-15>", line 2, in step
>   File "/opt/miniconda3/lib/python3.7/site-packages/gssapi/_utils.py", line 167, in check_last_err
>     return func(self, *args, **kwargs)
>   File "</opt/miniconda3/lib/python3.7/site-packages/decorator.py:decorator-gen-5>", line 2, in step
>   File "/opt/miniconda3/lib/python3.7/site-packages/gssapi/_utils.py", line 127, in catch_and_return_token
>     return func(self, *args, **kwargs)
>   File "/opt/miniconda3/lib/python3.7/site-packages/gssapi/sec_contexts.py", line 521, in step
>     return self._initiator_step(token=token)
>   File "/opt/miniconda3/lib/python3.7/site-packages/gssapi/sec_contexts.py", line 542, in _initiator_step
>     token)
>   File "gssapi/raw/sec_contexts.pyx", line 244, in gssapi.raw.sec_contexts.init_sec_context
> gssapi.raw.misc.GSSError: Major (851968): Unspecified GSS failure.  Minor code may provide more information, Minor (2529639053): Matching credential not found
> ERROR:requests_gssapi.gssapi_:Major (851968): Unspecified GSS failure.  Minor code may provide more information, Minor (2529639053): Matching credential not found
> Traceback (most recent call last):
>   File "/opt/miniconda3/lib/python3.7/site-packages/requests_gssapi/gssapi_.py", line 150, in generate_request_header
>     _negotiate_value(response))
>   File "</opt/miniconda3/lib/python3.7/site-packages/decorator.py:decorator-gen-15>", line 2, in step
>   File "/opt/miniconda3/lib/python3.7/site-packages/gssapi/_utils.py", line 167, in check_last_err
>     return func(self, *args, **kwargs)
>   File "</opt/miniconda3/lib/python3.7/site-packages/decorator.py:decorator-gen-5>", line 2, in step
>   File "/opt/miniconda3/lib/python3.7/site-packages/gssapi/_utils.py", line 127, in catch_and_return_token
>     return func(self, *args, **kwargs)
>   File "/opt/miniconda3/lib/python3.7/site-packages/gssapi/sec_contexts.py", line 521, in step
>     return self._initiator_step(token=token)
>   File "/opt/miniconda3/lib/python3.7/site-packages/gssapi/sec_contexts.py", line 542, in _initiator_step
>     token)
>   File "gssapi/raw/sec_contexts.pyx", line 244, in gssapi.raw.sec_contexts.init_sec_context
> gssapi.raw.misc.GSSError: Major (851968): Unspecified GSS failure.  Minor code may provide more information, Minor (2529639053): Matching credential not found
> DEBUG:requests_gssapi.gssapi_:handle_401(): returning <Response [401]>
> DEBUG:requests_gssapi.gssapi_:handle_response(): returning <Response [401]>
> DEBUG:requests_gssapi.gssapi_:handle_response() has seen 1 401 responses
> DEBUG:requests_gssapi.gssapi_:handle_response(): returning 401 <Response [401]>
> MYUSER@EXAMPLE.NET
> 32448
> {<OID 1.2.840.113554.1.2.2>}
> initiate
> <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
> <html><head>
> <title>401 Unauthorized</title>
> </head><body>
> <h1>Unauthorized</h1>
> <p>This server could not verify that you
> are authorized to access the document
> requested.  Either you supplied the wrong
> credentials (e.g., bad password), or your
> browser doesn't understand how to supply
> the credentials required.</p>
> </body></html>
> 
> INFO:app:r.text
> [11512] 1562863977.545786: Destroying ccache MEMORY:jDP7v1E
> INFO:werkzeug:10.16.70.114 - - [11/Jul/2019 18:52:57] "GET / HTTP/1.1" 200 -
> 
>

**Comment by jembishop** [comment]
Hi sorry for the delay, I have been testing the delegation out on this and it works on my machine.  I am guessing there is something wrong about how this is set up in the kdc but I am not sure.

**Comment by ptulpen** [comment]
Hello, well I was away as well 

What I see is that when I have no current ticket at all when starting(due to using kdestroy) , I get 

gssapi.raw.misc.GSSError: Major (851968): Unspecified GSS failure.  Minor code may provide more information, Minor (2529638936): Preauthentication failed

So is seems to using the ticket from local cache, not from the keytab.

**Comment by jembishop** [comment]
Yeah I am not sure what this issue is about I am afraid. I would reccomend asking someone more knowledgable about delegation than me if you haven't already.

**Comment by hackermandh** [comment]
The error:

`gssapi.raw.misc.GSSError: Major (851968): Unspecified GSS failure.  Minor code may provide more information, Minor (2529638919): Server not found in Kerberos database`

My solution:

Reduce the version of `kafka-python` until your code works again (or just rollback `poetry.lock` to the version version) - `gssapi` is used by `kafka-python`.

I had run a `poetry update`, which updated the version of `kafka-python`, but one of the latest (minor) versions broke something. I can't be bothered to figure out which one it was, since we're jumping ship off of Kafka soon anyway.

PS: Am only posting this here because this was one of the first results when searching for this exact error, in case I get the same error at a later date, or in case someone else finds this page in search for a solution of their Python code 😅
